util/argo/audit_logger.go

```diff
@@ -0,0 +1,104 @@
+package argo
+
+import (
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"fmt"
+	"time"
+
+	"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+type AuditLogger struct {
+	kIf       kubernetes.Interface
+	component string
+	ns        string
+}
+
+type EventInfo struct {
+	Type   string
+	Reason string
+}
+
+type ObjectRef struct {
+	Name            string
+	Namespace       string
+	ResourceVersion string
+	UID             types.UID
+}
+
+const (
+	EventReasonWorkflowRunning  = "WorkflowRunning"
+	EventReasonWorkflowSucceded = "WorkflowSucceeded"
+	EventReasonWorkflowFailed   = "WorkflowFailed"
+	EventReasonWorkflowTimedOut = "WorkflowTimedOut"
+)
+
+func (l *AuditLogger) logEvent(objMeta ObjectRef, gvk schema.GroupVersionKind, info EventInfo, message string, logFields map[string]string) {
+	logCtx := log.WithFields(log.Fields{
+		"type":   info.Type,
+		"reason": info.Reason,
+	})
+	for field, val := range logFields {
+		logCtx = logCtx.WithField(field, val)
+	}
+
+	switch gvk.Kind {
+	case "Workflow":
+		logCtx = logCtx.WithField("workflow", objMeta.Name)
+	default:
+		logCtx = logCtx.WithField("name", objMeta.Name)
+	}
+	t := metav1.Time{Time: time.Now()}
+	event := v1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%v.%x", objMeta.Name, t.UnixNano()),
+		},
+		Source: v1.EventSource{
+			Component: l.component,
+		},
+		InvolvedObject: v1.ObjectReference{
+			Kind:            gvk.Kind,
+			Name:            objMeta.Name,
+			Namespace:       objMeta.Namespace,
+			ResourceVersion: objMeta.ResourceVersion,
+			APIVersion:      gvk.Version,
+			UID:             objMeta.UID,
+		},
+		FirstTimestamp: t,
+		LastTimestamp:  t,
+		Count:          1,
+		Message:        message,
+		Type:           info.Type,
+		Reason:         info.Reason,
+	}
+	logCtx.Info(message)
```

Comment:
should probably be structured? so we can write Splunk (or other logging facility) queries e.g `log.WithField("event", event)`. Should this be debug?

Reply:
I'd love a structured log (we actually use Honeycomb), i'll think about it

Reply:
I met the Honeycomb CEO earlier this year - really interesting person.

Reply:
done